File: android/src/main/kotlin/re/notifica/flutter/models/CoreModels.kt
package re.notifica.flutter.models

import re.notifica.flutter.NotificarePlugin
import re.notifica.models.*
import java.util.*

internal typealias FlutterMap = Map<String, Any?>

internal fun NotificareApplication.toFlutterMap(): FlutterMap = mapOf(
    "id" to id,
    "name" to name,
    "category" to category,
    "services" to services,
    "inboxConfig" to inboxConfig?.toFlutterMap(),
    "regionConfig" to regionConfig?.toFlutterMap(),
    "userDataFields" to userDataFields.map { it.toFlutterMap() },
    "actionCategories" to actionCategories.map { it.toFlutterMap() },
)

internal fun NotificareApplication.InboxConfig.toFlutterMap(): FlutterMap = mapOf(
    "useInbox" to useInbox,
    "autoBadge" to autoBadge,
)

internal fun NotificareApplication.RegionConfig.toFlutterMap(): FlutterMap = mapOf(
    "proximityUUID" to proximityUUID,
)

internal fun NotificareApplication.UserDataField.toFlutterMap(): FlutterMap = mapOf(
    "type" to type,
    "key" to key,
    "label" to label,
)

internal fun NotificareApplication.ActionCategory.toFlutterMap(): FlutterMap = mapOf(
    "type" to type,
    "name" to name,
    // TODO add remaining properties
)

internal fun NotificareDevice.toFlutterMap(): FlutterMap = mapOf(
    "id" to id,
    "userId" to userId,
    "userName" to userName,
    "timeZoneOffset" to timeZoneOffset,
    "osVersion" to osVersion,
    "sdkVersion" to sdkVersion,
    "appVersion" to appVersion,
    "deviceString" to deviceString,
    "language" to language,
    "region" to region,
    "transport" to NotificarePlugin.moshi.adapter(NotificareTransport::class.java).toJsonValue(transport),
    "dnd" to dnd?.toFlutterMap(),
    "userData" to userData,
    "lastRegistered" to NotificarePlugin.moshi.adapter(Date::class.java).toJsonValue(lastRegistered),
)

internal fun NotificareDoNotDisturb.toFlutterMap(): FlutterMap = mapOf(
    "start" to start.format(),
    "end" to end.format(),
)

internal fun NotificareDoNotDisturb(json: FlutterMap): NotificareDoNotDisturb {
    return NotificareDoNotDisturb(
        start = NotificareTime(json["start"] as String),
        end = NotificareTime(json["end"] as String),
    )
}

File: notificare-core/ios/Classes/Models/CoreModels.swift
//
//  CoreModels.swift
//  notificare
//
//  Created by Helder Pinhal on 03/12/2020.
//

import Foundation
import NotificareKit

extension NotificareApplication {
    
    func toDictionary() -> FlutterDictionary {
        return [
            "id": id,
            "name": name,
            "category": category,
            "services": services,
            "inboxConfig": inboxConfig?.toDictionary(),
            "regionConfig": regionConfig?.toDictionary(),
            "userDataFields": userDataFields.map { $0.toDictionary() },
            "actionCategories": actionCategories.map { $0.toDictionary() },
        ]
    }
}

extension NotificareApplication.InboxConfig {
    
    func toDictionary() -> FlutterDictionary {
        return [
            "useInbox": useInbox,
            "autoBadge": autoBadge,
        ]
    }
}

extension NotificareApplication.RegionConfig {
    
    func toDictionary() -> FlutterDictionary {
        return [
            "proximityUUID": proximityUUID,
        ]
    }
}

extension NotificareApplication.UserDataField {
    
    func toDictionary() -> FlutterDictionary {
        return [
            "type": type,
            "key": key,
            "label": label,
        ]
    }
}

extension NotificareApplication.ActionCategory {
    
    func toDictionary() -> FlutterDictionary {
        return [
            "type": type,
            "name": name,
            // TODO add remaining properties
        ]
    }
}

extension NotificareDevice {

    func toDictionary() -> FlutterDictionary {
        return [
            "id": self.id,
            "userId": self.userId,
            "userName": self.userName,
            "timeZoneOffset": self.timeZoneOffset,
            "osVersion": self.osVersion,
            "sdkVersion": self.sdkVersion,
            "appVersion": self.appVersion,
            "deviceString": self.deviceString,
            "language": self.language,
            "region": self.region,
            "transport": self.transport.rawValue,
            "dnd": self.dnd?.toDictionary(),
            "userData": self.userData,
            "lastRegistered": SwiftNotificarePlugin.dateFormatter.string(from: self.lastRegistered),
        ]
    }
}

extension NotificareDoNotDisturb {

    init(dictionary: FlutterDictionary) throws {
        self.init(
            start: try NotificareTime(string: dictionary["start"] as! String),
            end: try NotificareTime(string: dictionary["end"] as! String)
        )
    }

    func toDictionary() -> [String: Any?] {
        return [
            "start": self.start.format(),
            "end": self.end.format(),
        ]
    }
}
